### ISSUE TITLE
Use somltcp

### ISSUE THREAD

**Issue by ibigbug** [Watfaq/clash-rs#208]

**Comment by cavivie** [comment]
@ibigbug I think I might be of some help here, as I recently stabilized the netstack-smoltcp crate.

**Comment by cavivie** [comment]
Well, I misunderstood you. It seems that you mean tcp/udp.

**Comment by ibigbug** [comment]
yeah that's waht i meant

**Comment by ibigbug** [comment]
@cavivie were you planning to  creat a PR for this?

**Comment by cavivie** [comment]
I'd love to, but before that, can you add more about tun in ss, including unresolved problems and known problems and goals.

**Comment by ibigbug** [comment]
I think the main goal is to replace the lwip w smoltcp, for performance boost, see https://github.com/Watfaq/netstack-smoltcp/blob/main/benchmark.md

And adopting pure rust impls over foreign C bindings. 

This is only for the tun input, not much todo with ss. 

We already use smoltcp in the user space wirefuard, it consolidates the deps which is another good reason too. 

And thanks for you work.

**Comment by cavivie** [comment]
Thank you for your description. I think I will take time to integrate it recently.

**Comment by VendettaReborn** [comment]
@cavivie @ibigbug i added some soem auto detect interface feature for rust-tun, and here is the example code: https://github.com/VendettaReborn/netstack-smoltcp/blob/f89b9e9b2e56a6abf1b2be9290e4cf63c7be3e11/examples/forward.rs#L142 . Basically, it's just a code version of route commands. And i've tested it on linux,macos & windows.

Shall we extract the logic of auto-detect interface, auto setup routes into a independent crate?  just like sing-tun?

**Comment by Itsusinn** [comment]
> Shall we extract the logic of auto-detect interface, auto setup routes into a independent crate?  just like sing-tun? 
i would like call it `watfaq-tun` : )

**Comment by VendettaReborn** [comment]
any updates?

**Comment by Itsusinn** [comment]
> any updates?

最近挺忙的，大概到六月初才有时间

**Comment by cavivie** [comment]
Ah, ah, first of all, I'm sorry for the reason why I've been busy recently. I want to disassemble the current steps here:
1. 
> Shall we extract the logic of auto-detect interface, auto setup routes into an independent crate? just like sing-tun?

Are there any disassembled independent clates here? I want to do some work based on this.

2.
BTW, netstack-smoltcp may have some things that need to be changed. Can I think of anything else you need to add besides some refactoring of the udp part?

**Comment by VendettaReborn** [comment]
hh, it's ok, and hope you can rejoin the discussion whenever you're ready. 

for the first point, for example, the route's  auto-configuration can be done via programming, here is a WIP version: (https://github.com/VendettaReborn/netstack-smoltcp/blob/auto-detect-interface/src/utils.rs), it can works in netstack-smoltcp's example, but need some extra work before adding it into clash-rs. 

What's your idea?

**Comment by ibigbug** [comment]
> i would like call it watfaq-tun : )

Or wontun 😄

**Comment by VendettaReborn** [comment]
i've done some extra jobs on it:
1. [forward_lwip](https://github.com/VendettaReborn/netstack-smoltcp/blob/auto-detect-interface/examples/forward_lwip.rs) && [forward_smoltcp](https://github.com/VendettaReborn/netstack-smoltcp/blob/auto-detect-interface/examples/forward_smoltcp.rs). I compared smoltcp with lwip on their basic proxy mechanism, and found that the netstack-smoltcp may not work well on ipv6 (lwip is good)
2. some [work](https://github.com/VendettaReborn/netstack-smoltcp/blob/a8b3e3bc19d87d275c89214e854e758c5c43c096/src/utils.rs#L84) on adding ipv6 addr for tun device, which is missed in rust-tun
3.  [watfaq-tun](https://github.com/VendettaReborn/watfaq-tun). code equivalent for `ip route` & 'ip rule' [commands](https://github.com/automesh-network/netstack-smoltcp/blob/main/scripts/route-linux.sh). But with supports of ipv6. Now works on linux&macos.

and watfaq-tun may need some reviews and comments. We can make it more robust and user-friendly before putting into usage in clash-rs. 

cc @ibigbug @cavivie @Itsusinn

**Comment by ibigbug** [comment]
Awesome work! Thanks! Will take a look

**Comment by ibigbug** [comment]
@VendettaReborn i think we can start integrating your tun with https://github.com/Watfaq/clash-rs/issues/396 and improve it progressively, thoughts?

**Comment by Itsusinn** [comment]
> [watfaq-tun](https://github.com/VendettaReborn/watfaq-tun). code equivalent for `ip route` & 'ip rule' [commands](https://github.com/automesh-network/netstack-smoltcp/blob/main/scripts/route-linux.sh). But with supports of ipv6. Now works on linux&macos.

I dont think we need net_route crate(at least on linux). It doesn't meet my requirement of impl auto-route(eg. no fwmark support).

**Comment by VendettaReborn** [comment]
of course, global routing on linux is easy, your impl of fwmark is surely sufficient, but what about macos and windows? I don't expect most of our users to use linux.

**Comment by Itsusinn** [comment]
> of course, global routing on linux is easy, your impl of fwmark is surely sufficient, but what about macos and windows? I don't expect most of our users to use linux.

I could write my route impl on windows. But indeed i cant do the same for macos xD. So i would like take net-route as a platform-specific dep.

**Comment by VendettaReborn** [comment]
i've wrote another crate [ipconfig2](https://github.com/hsqStephenZhang/ipconfig2), i can be integrated into net-route as well, after that, the pc platforms shall all be supported

**Comment by cavivie** [comment]
> [forward_lwip](https://github.com/VendettaReborn/netstack-smoltcp/blob/auto-detect-interface/examples/forward_lwip.rs) && [forward_smoltcp](https://github.com/VendettaReborn/netstack-smoltcp/blob/auto-detect-interface/examples/forward_smoltcp.rs). I compared smoltcp with lwip on their basic proxy mechanism, and found that the netstack-smoltcp may not work well on ipv6 (lwip is good)

I think we may be further along, since ipv6 local routing is supported after [this MR](https://github.com/smoltcp-rs/smoltcp/pull/900), I have successfully tested ipv6 ping in `netstack-smoltcp` using the latest main branch after this MR: https://github.com/smoltcp-rs/smoltcp/pull/900

**Comment by cavivie** [comment]
For now, this means we can further work on IPv6 code. Nonetheless, this also seems to mean that the support of the ipv6 part of the smoltcp crate is not perfect enough. We may need to test more to find the problem. I am not sure what specific aspects we need to test for IPv6. Can you give me some advice?

**Comment by VendettaReborn** [comment]
for the outbound handlers, we have the docker test suites, which will cover tcp&udp. For tun, we haven't figure out some good ways to test it on both ipv4 & ipv6. here is a brief thought: https://github.com/Watfaq/clash-rs/pull/362#issuecomment-2114260969

**Comment by cavivie** [comment]
Here is a pic that some of simple tcp tests shared. `rust-tun` does not currently support ipv6, I want to do something about `rust-tun`'s shortcomings in ipv6 in the near future to avoid us do some hacky things about ipv6 for tun device at outside of `rust-tun` crate. What do you think?

![image](https://github.com/Watfaq/clash-rs/assets/103884097/71b4a8b0-2cf2-4f66-bec8-153351895f64)

**Comment by cavivie** [comment]
> for the outbound handlers, we have the docker test suites, which will cover tcp&udp. For tun, we haven't figure out some good ways to test it on both ipv4 & ipv6. here is a brief thought: [#362 (comment)](https://github.com/Watfaq/clash-rs/pull/362#issuecomment-2114260969)
> 
> @cavivie thoughts?

Absolutely agree, usually routing is definitely an important point that hinders testing.

**Comment by VendettaReborn** [comment]
yeah, i agree. rust-tun hasn't supported ipv6 yet, we can extend the ability and some other feature(like routing rules setup) in a separate crate.

> Here is a pic that some of simple tcp tests shared. `rust-tun` does not currently support ipv6, I want to do something about `rust-tun`'s shortcomings in ipv6 in the near future to avoid us do some hacky things about ipv6 for tun device at outside of `rust-tun` crate. What do you think?
> 
> ![image](https://private-user-images.githubusercontent.com/103884097/333792705-71b4a8b0-2cf2-4f66-bec8-153351895f64.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY2MzUxNzAsIm5iZiI6MTcxNjYzNDg3MCwicGF0aCI6Ii8xMDM4ODQwOTcvMzMzNzkyNzA1LTcxYjRhOGIwLTJjZjItNGY2Ni1iZWM4LTE1MzM1MTg5NWY2NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyNVQxMTAxMTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yODg3NzViYTAzZmQzMDA0ZmVmOTM4MDI5NWVkZjRhM2U0NDFjNjUwNjExMDNmZmM0NjU3N2YwMmMyNDczN2NhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.SHf0W08u9olSAbbJTyhUgj18ByuWVXrwIIqTrCDqDag)

**Comment by cavivie** [comment]
>a separate crate
What I mean is that we may need to use the  `rust-tun` patch first, and then merge some patches about ipv6 into `rust-tun` upstream if possible.

**Comment by VendettaReborn** [comment]
oh, i see, agreed.

**Comment by cavivie** [comment]
To add tun feature, we want ` rust-tun` to behave like:
- Dual stack is enabled by default, that is, ipv4/ipv6 addresses are bound to the tun device at the same time, and then a method is provided to disable ipv6 or ipv4;
- IPv4 is enabled by default, IPv6 can be optionally added
- IPv6 is enabled by default, IPv4 can be optionally added

Obviously, I think the first two are more reliable, and in particular, I think the second point is more needed. What do you think?

**Comment by VendettaReborn** [comment]
that would be great

**Comment by cavivie** [comment]
After some time testing tun2 and various netstacks, I found that in terms of throughput:
- Throughput (nix write): lwip > smoltcp > ipstack
- Throughput (win write): lwip > smoltcp > ipstack
- Throughput (nix read): smoltcp = lwip > ipstack
- Throughput (win read): smoltcp > lwip > ipstack

In terms of stability, ipstack currently performs the worst, and even the throughput is unacceptable on win, about 0.2 to 0.3 Gbps., while the VPN scenario focuses more on netstack's read performance, followed by write performance, and finally stability.

The above data is only a personal test on a specific device. Later, I will publish a test routine for you to test and compare.

**Comment by cavivie** [comment]
rust run pef test: https://github.com/automesh-network/rust-tun-perf.git

**Comment by ibigbug** [comment]
Thanks for sharing. 

To me I think it's still worth adopting smoltcp even with similar bar of performance to lwip for simplicity and pure rust impl.
